subsys/fs/nffs_fs.c

```diff
@@ -0,0 +1,540 @@
+		  CONFIG_FILE_SYSTEM_NFFS_NUM_BLOCKS,	4);
+K_MEM_SLAB_DEFINE(nffs_cache_inode_pool,	sizeof(struct nffs_cache_inode),
+		  CONFIG_FILE_SYSTEM_NFFS_NUM_CACHE_INODES,	4);
+K_MEM_SLAB_DEFINE(nffs_cache_block_pool,	sizeof(struct nffs_cache_block),
+		  CONFIG_FILE_SYSTEM_NFFS_NUM_CACHE_BLOCKS,	4);
+
+static int translate_error(int error)
+{
+	switch (error) {
+	case FS_EOK:
+		return 0;
+	case FS_ECORRUPT:
+	case FS_EHW:
+		return -EIO;
+	case FS_EOFFSET:
+	case FS_EINVAL:
+		return -EINVAL;
+	case FS_ENOMEM:
+		return -ENOMEM;
+	case FS_ENOENT:
+		return -ENOENT;
+	case FS_EEMPTY:
+		return -ENODEV;
+	case FS_EFULL:
+		return -ENOSPC;
+	case FS_EUNEXP:
+	case FS_EOS:
+		return -EIO;
+	case FS_EEXIST:
+		return -EEXIST;
+	case FS_EACCESS:
+		return -EACCES;
+	case FS_EUNINIT:
+		return -EIO;
+	}
+
+	return -EIO;
+}
+
+int nffs_os_mempool_init(void)
+{
+	/*
+	 * Just reinitialize slabs here - this is what original implementation
+	 * does. We assume all references to previously allocated blocks, if
+	 * any, are invalidated in NFFS code already.
+	 */
+
+	k_mem_slab_init(&nffs_file_pool, _k_mem_slab_buf_nffs_file_pool,
+			sizeof(struct nffs_file),
+			CONFIG_FILE_SYSTEM_NFFS_NUM_FILES);
+	k_mem_slab_init(&nffs_dir_pool, _k_mem_slab_buf_nffs_dir_pool,
+			sizeof(struct nffs_dir),
+			CONFIG_FILE_SYSTEM_NFFS_NUM_DIRS);
+	k_mem_slab_init(&nffs_inode_entry_pool,
+			_k_mem_slab_buf_nffs_inode_entry_pool,
+			sizeof(struct nffs_inode_entry),
+			CONFIG_FILE_SYSTEM_NFFS_NUM_INODES);
+	k_mem_slab_init(&nffs_block_entry_pool,
+			_k_mem_slab_buf_nffs_block_entry_pool,
+			sizeof(struct nffs_hash_entry),
+			CONFIG_FILE_SYSTEM_NFFS_NUM_BLOCKS);
+	k_mem_slab_init(&nffs_cache_inode_pool,
+			_k_mem_slab_buf_nffs_cache_inode_pool,
+			sizeof(struct nffs_cache_inode),
+			CONFIG_FILE_SYSTEM_NFFS_NUM_CACHE_INODES);
+	k_mem_slab_init(&nffs_cache_block_pool,
+			_k_mem_slab_buf_nffs_cache_block_pool,
+			sizeof(struct nffs_cache_block),
+			CONFIG_FILE_SYSTEM_NFFS_NUM_CACHE_BLOCKS);
+
+	return 0;
+}
+
+void *nffs_os_mempool_get(nffs_os_mempool_t *pool)
+{
+	int rc;
+	void *ptr;
+
+	rc = k_mem_slab_alloc(pool, &ptr, K_NO_WAIT);
+	if (rc) {
+		ptr = NULL;
+	}
+
+	return ptr;
+}
+
+int nffs_os_mempool_free(nffs_os_mempool_t *pool, void *block)
+{
+	k_mem_slab_free(pool, &block);
+
+	return 0;
+}
+
+int nffs_os_flash_read(u8_t id, u32_t address, void *dst, u32_t num_bytes)
+{
+	int rc;
+
+	rc = flash_read(flash_dev, address, dst, num_bytes);
+
+	return rc;
+}
+
+int nffs_os_flash_write(u8_t id, u32_t address, const void *src,
+			u32_t num_bytes)
+{
+	int rc;
+
+	rc = flash_write_protection_set(flash_dev, 0);
+	if (rc) {
+		return rc;
+	}
+
+	rc = flash_write(flash_dev, address, src, num_bytes);
+
+	return rc;
+}
+
+int nffs_os_flash_erase(u8_t id, u32_t address, u32_t num_bytes)
+{
+	int rc;
+
+	rc = flash_write_protection_set(flash_dev, 0);
+	if (rc) {
+		return rc;
+	}
+
+	rc = flash_erase(flash_dev, address, num_bytes);
+
```

Comment:
After the write flash may be leaved unprotected for some devic... see coment above.